File: tests/workflow/consumer/rust/src/stat.rs
use console::style;
use std::collections::HashMap;

#[allow(non_upper_case_globals)]
mod expectations {
    // Expected broadcasts ============================
    pub const GroupAStructA: usize = 1;
    pub const GroupAStructB: usize = 1;
    pub const GroupBGroupCStructA: usize = 1;
    pub const GroupBGroupCStructB: usize = 2;
    pub const GroupBStructA: usize = 1;
    pub const StructA: usize = 1;
    pub const StructB: usize = 2;
    pub const StructC: usize = 2;
    pub const StructD: usize = 3;
    pub const StructF: usize = 2;
    pub const StructJ: usize = 2;
    // Expected broadcasts ============================
    pub const TriggerBeacons: usize = 1;
    pub const FinishConsumerTestBroadcast: usize = 1;
    pub const Connected: usize = 1;
    pub const Disconnected: usize = 1;
    pub const Shutdown: usize = 1;
    pub const TestRequestGroupAStructA: usize = 3;
    pub const TestRequestGroupAStructB: usize = 3;
    pub const TestRequestGroupBGroupCStructA: usize = 2;
    pub const TestRequestGroupBGroupCStructB: usize = 4;
    pub const TestRequestGroupBStructA: usize = 3;
    pub const TestRequestStructA: usize = 4;
    pub const TestRequestStructC: usize = 4;
    pub const TestRequestStructD: usize = 2;
    pub const TestRequestStructF: usize = 2;
    pub const TestRequestStructRmpty: usize = 2;
}

#[derive(PartialEq, Hash, PartialOrd)]
pub enum Alias {
    GroupAStructA,
    GroupAStructB,
    GroupBGroupCStructA,
    GroupBGroupCStructB,
    GroupBStructA,
    StructA,
    StructB,
    StructC,
    StructD,
    StructF,
    StructJ,
    TriggerBeacons,
    FinishConsumerTestBroadcast,
    Connected,
    Disconnected,
    Shutdown,
    TestRequestGroupAStructA,
    TestRequestGroupAStructB,
    TestRequestGroupBGroupCStructA,
    TestRequestGroupBGroupCStructB,
    TestRequestGroupBStructA,
    TestRequestStructA,
    TestRequestStructC,
    TestRequestStructD,
    TestRequestStructF,
    TestRequestStructRmpty,
}

impl Eq for Alias {}

impl std::fmt::Display for Alias {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::GroupAStructA => write!(f, "GroupAStructA"),
            Self::GroupAStructB => write!(f, "GroupAStructB"),
            Self::GroupBGroupCStructA => write!(f, "GroupBGroupCStructA"),
            Self::GroupBGroupCStructB => write!(f, "GroupBGroupCStructB"),
            Self::GroupBStructA => write!(f, "GroupBStructA"),
            Self::StructA => write!(f, "StructA"),
            Self::StructB => write!(f, "StructB"),
            Self::StructC => write!(f, "StructC"),
            Self::StructD => write!(f, "StructD"),
            Self::StructF => write!(f, "StructF"),
            Self::StructJ => write!(f, "StructJ"),
            Self::TriggerBeacons => write!(f, "TriggerBeacons"),
            Self::FinishConsumerTestBroadcast => write!(f, "FinishConsumerTestBroadcast"),
            Self::Connected => write!(f, "Connected"),
            Self::Disconnected => write!(f, "Disconnected"),
            Self::Shutdown => write!(f, "Shutdown"),
            Self::TestRequestGroupAStructA => write!(f, "TestRequestGroupAStructA"),
            Self::TestRequestGroupAStructB => write!(f, "TestRequestGroupAStructB"),
            Self::TestRequestGroupBGroupCStructA => write!(f, "TestRequestGroupBGroupCStructA"),
            Self::TestRequestGroupBGroupCStructB => write!(f, "TestRequestGroupBGroupCStructB"),
            Self::TestRequestGroupBStructA => write!(f, "TestRequestGroupBStructA"),
            Self::TestRequestStructA => write!(f, "TestRequestStructA"),
            Self::TestRequestStructC => write!(f, "TestRequestStructC"),
            Self::TestRequestStructD => write!(f, "TestRequestStructD"),
            Self::TestRequestStructF => write!(f, "TestRequestStructF"),
            Self::TestRequestStructRmpty => write!(f, "TestRequestStructRmpty"),
        }
    }
}

pub enum StatEvent {
    Inc(Alias),
    ConsumerDone,
}

pub struct Stat {
    connections: usize,
    done: usize,
    pub tests: HashMap<Alias, (usize, usize)>,
}

impl Stat {
    pub fn new(connections: usize) -> Self {
        let mut tests = HashMap::new();
        tests.insert(
            Alias::GroupAStructA,
            (0, connections * expectations::GroupAStructA),
        );
        tests.insert(
            Alias::GroupAStructB,
            (0, connections * expectations::GroupAStructB),
        );
        tests.insert(
            Alias::GroupBGroupCStructA,
            (0, connections * expectations::GroupBGroupCStructA),
        );
        tests.insert(
            Alias::GroupBGroupCStructB,
            (0, connections * expectations::GroupBGroupCStructB),
        );
        tests.insert(
            Alias::GroupBStructA,
            (0, connections * expectations::GroupBStructA),
        );
        tests.insert(Alias::StructA, (0, connections * expectations::StructA));
        tests.insert(Alias::StructB, (0, connections * expectations::StructB));
        tests.insert(Alias::StructC, (0, connections * expectations::StructC));
        tests.insert(Alias::StructD, (0, connections * expectations::StructD));
        tests.insert(Alias::StructF, (0, connections * expectations::StructF));
        tests.insert(Alias::StructJ, (0, connections * expectations::StructJ));
        tests.insert(
            Alias::TriggerBeacons,
            (0, connections * expectations::TriggerBeacons),
        );
        tests.insert(
            Alias::FinishConsumerTestBroadcast,
            (0, connections * expectations::FinishConsumerTestBroadcast),
        );
        tests.insert(Alias::Connected, (0, connections * expectations::Connected));
        tests.insert(
            Alias::Disconnected,
            (0, connections * expectations::Disconnected),
        );
        tests.insert(Alias::Shutdown, (0, connections * expectations::Shutdown));
        tests.insert(
            Alias::TestRequestGroupAStructA,
            (0, connections * expectations::TestRequestGroupAStructA),
        );
        tests.insert(
            Alias::TestRequestGroupAStructB,
            (0, connections * expectations::TestRequestGroupAStructB),
        );
        tests.insert(
            Alias::TestRequestGroupBGroupCStructA,
            (
                0,
                connections * expectations::TestRequestGroupBGroupCStructA,
            ),
        );
        tests.insert(
            Alias::TestRequestGroupBGroupCStructB,
            (
                0,
                connections * expectations::TestRequestGroupBGroupCStructB,
            ),
        );
        tests.insert(
            Alias::TestRequestGroupBStructA,
            (0, connections * expectations::TestRequestGroupBStructA),
        );
        tests.insert(
            Alias::TestRequestStructA,
            (0, connections * expectations::TestRequestStructA),
        );
        tests.insert(
            Alias::TestRequestStructC,
            (0, connections * expectations::TestRequestStructC),
        );
        tests.insert(
            Alias::TestRequestStructD,
            (0, connections * expectations::TestRequestStructD),
        );
        tests.insert(
            Alias::TestRequestStructF,
            (0, connections * expectations::TestRequestStructF),
        );
        tests.insert(
            Alias::TestRequestStructRmpty,
            (0, connections * expectations::TestRequestStructRmpty),
        );
        Self {
            connections,
            tests,
            done: 0,
        }
    }

    pub fn apply(&mut self, event: StatEvent) {
        match event {
            StatEvent::Inc(alias) => {
                if let Some((current, _)) = self.tests.get_mut(&alias) {
                    *current += 1;
                }
            }
            StatEvent::ConsumerDone => {
                self.done += 1;
            }
        }
    }
}

impl std::fmt::Display for Stat {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut output = format!("{}\n", "=".repeat(70));
        let mut tests = vec![];
        for (alias, (current, expectation)) in &self.tests {
            tests.push((alias.to_string(), current, expectation));
        }
        tests.sort();
        for (alias, current, expectation) in &tests {
            let mut alias = alias.to_string();
            let filler = 60 - alias.len();
            alias = format!(
                "{}{}",
                alias,
                ".".repeat(if filler > 0 { filler } else { 0 })
            );
            output = format!(
                "{} {} {}: {}/{}\n",
                output,
                if current == expectation {
                    style("☑").bold().green().dim()
                } else {
                    style("☐").bold().red().dim()
                },
                style(format!("[{}]", alias)).bold().dim(),
                if current == expectation {
                    style(format!("{}", current)).dim()
                } else {
                    style(format!("{}", current)).bold().red().dim()
                },
                if current == expectation {
                    style(format!("{}", expectation)).dim()
                } else {
                    style(format!("{}", expectation)).bold().red().dim()
                }
            );
        }
        output = format!("{}{}\n", output, "=".repeat(70));
        write!(f, "{}", output)
    }
}

File: tests/workflow/consumer/typescript/src/stat.ts
export const expectations = {
	BeaconA: 1,
	BeaconsBeaconA: 1,
	BeaconsBeaconB: 1,
	BeaconsSubBeaconA: 1,
	GroupAStructA: 1,
	GroupAStructB: 1,
	GroupBGroupCStructA: 1,
	GroupBGroupCStructB: 2,
	GroupBStructA: 1,
	StructA: 1,
	StructB: 2,
	StructC: 2,
	StructD: 3,
	StructF: 2,
	StructJ: 2,
	TriggerBeacons: 1,
	FinishConsumerTestBroadcast: 1,
	Connected: 1,
	Disconnected: 1,
	Error: 0,
};

export enum Alias {
	BeaconA = "BeaconA",
	BeaconsBeaconA = "BeaconsBeaconA",
	BeaconsBeaconB = "BeaconsBeaconB",
	BeaconsSubBeaconA = "BeaconsSubBeaconA",
	GroupAStructA = "GroupAStructA",
	GroupAStructB = "GroupAStructB",
	GroupBGroupCStructA = "GroupBGroupCStructA",
	GroupBGroupCStructB = "GroupBGroupCStructB",
	GroupBStructA = "GroupBStructA",
	StructA = "StructA",
	StructB = "StructB",
	StructC = "StructC",
	StructD = "StructD",
	StructF = "StructF",
	StructJ = "StructJ",
	TriggerBeacons = "TriggerBeacons",
	FinishConsumerTestBroadcast = "FinishConsumerTestBroadcast",
	Connected = "Connected",
	Disconnected = "Disconnected",
	Error = "Error",
}

const all: Alias[] = [
	Alias.BeaconA,
	Alias.BeaconsBeaconA,
	Alias.BeaconsBeaconB,
	Alias.BeaconsSubBeaconA,
	Alias.GroupAStructA,
	Alias.GroupAStructB,
	Alias.GroupBGroupCStructA,
	Alias.GroupBGroupCStructB,
	Alias.GroupBStructA,
	Alias.StructA,
	Alias.StructB,
	Alias.StructC,
	Alias.StructD,
	Alias.StructF,
	Alias.StructJ,
	Alias.TriggerBeacons,
	Alias.FinishConsumerTestBroadcast,
	Alias.Connected,
	Alias.Disconnected,
	Alias.Error,
];

export interface TestData {
	done: number;
	expectation: number;
}

type Handler = () => void;

export class Stat {
	static MIN_BROADCAST: number = 19;
	private _tests: Map<Alias, TestData> = new Map();
	private _indexes: Map<Alias, number> = new Map();
	private _finishHandler: Handler | undefined;
	private _wating: {
		broadcast: boolean;
		beacons: boolean;
		finish: boolean;
	} = {
		broadcast: false,
		beacons: false,
		finish: false,
	};

	constructor(noExpectations?: boolean) {
		all.forEach((alias) => {
			this._indexes.set(alias, 0);
		});
		this._tests.set(Alias.BeaconA, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.BeaconA,
		});
		this._tests.set(Alias.BeaconsBeaconA, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.BeaconsBeaconA,
		});
		this._tests.set(Alias.BeaconsBeaconB, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.BeaconsBeaconB,
		});
		this._tests.set(Alias.BeaconsSubBeaconA, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.BeaconsSubBeaconA,
		});
		this._tests.set(Alias.GroupAStructA, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.GroupAStructA,
		});
		this._tests.set(Alias.GroupAStructB, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.GroupAStructB,
		});
		this._tests.set(Alias.GroupBGroupCStructA, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.GroupBGroupCStructA,
		});
		this._tests.set(Alias.GroupBGroupCStructB, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.GroupBGroupCStructB,
		});
		this._tests.set(Alias.GroupBStructA, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.GroupBStructA,
		});
		this._tests.set(Alias.StructA, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.StructA,
		});
		this._tests.set(Alias.StructB, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.StructB,
		});
		this._tests.set(Alias.StructC, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.StructC,
		});
		this._tests.set(Alias.StructD, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.StructD,
		});
		this._tests.set(Alias.StructF, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.StructF,
		});
		this._tests.set(Alias.StructJ, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.StructJ,
		});
		this._tests.set(Alias.TriggerBeacons, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.TriggerBeacons,
		});
		this._tests.set(Alias.FinishConsumerTestBroadcast, {
			done: 0,
			expectation: noExpectations
				? 0
				: expectations.FinishConsumerTestBroadcast,
		});
		this._tests.set(Alias.Connected, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.Connected,
		});
		this._tests.set(Alias.Disconnected, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.Disconnected,
		});
		this._tests.set(Alias.Error, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.Error,
		});
	}

	public index(alias: Alias): number {
		const index = this._indexes.get(alias) as number;
		this._indexes.set(alias, index + 1);
		return index;
	}

	public case(alias: Alias) {
		const data = this._tests.get(alias) as TestData;
		data.done += 1;
		this._tests.set(alias, data);
		if (!this._wating.broadcast) {
			if (this.total() >= Stat.MIN_BROADCAST) {
				this._wating.broadcast = true;
				this._finish();
			}
		}
	}

	public total(): number {
		let total: number = 0;
		Array.from(this._tests.values()).forEach((data: TestData) => {
			total += data.done;
		});
		return total;
	}

	public onFinish(cb: Handler) {
		this._finishHandler = cb;
	}

	public beacons() {
		this._wating.beacons = true;
		this._finish();
	}

	public finish() {
		this._wating.finish = true;
		this._finish();
	}

	public merge(stat: Stat) {
		this._tests.forEach((data: TestData, alias: Alias) => {
			const income = stat.getCase(alias);
			data.done += income.done;
			data.expectation += income.expectation;
		});
	}

	public print() {
		const LEN: number = 70;
		this._tests.forEach((data: TestData, alias: Alias) => {
			const title = `${alias}`;
			const correct = data.done === data.expectation;
			console.log(
				`${correct ? "☑" : "☐"} ${title}${".".repeat(
					LEN - title.length
				)}: ${data.done} / ${data.expectation}`
			);
		});
	}

	public getCase(alias: Alias): TestData {
		return this._tests.get(alias) as TestData;
	}

	public getErrors(): string[] {
		const errors: string[] = [];
		this._tests.forEach((data: TestData, alias: Alias) => {
			if (data.done !== data.expectation) {
				errors.push(
					`Test for usecase "${alias}" is failed; ${data.done} / ${data.expectation}`
				);
			}
		});
		return errors;
	}

	private _finish() {
		if (
			this._wating.beacons &&
			this._wating.broadcast &&
			this._wating.finish
		) {
			this._finishHandler !== undefined && this._finishHandler();
			this._finishHandler = undefined;
		}
	}
}
